src/solvers/iterators.jl

```diff
@@ -0,0 +1,104 @@
+#
+# SweepIterator
+#
+
+mutable struct SweepIterator
+  sweep_kws
+  region_iter
+  which_sweep::Int
+end
+
+problem(S::SweepIterator) = problem(S.region_iter)
+
+Base.length(S::SweepIterator) = length(S.sweep_kws)
+
+function Base.iterate(S::SweepIterator, which=nothing)
+  if isnothing(which)
+    sweep_kws_state = iterate(S.sweep_kws)
+  else
+    sweep_kws_state = iterate(S.sweep_kws, which)
+  end
+  isnothing(sweep_kws_state) && return nothing
+  current_sweep_kws, next = sweep_kws_state
+
+  if !isnothing(which)
+    S.region_iter = region_iterator(
+      problem(S.region_iter); sweep=S.which_sweep, current_sweep_kws...
+    )
+  end
+  S.which_sweep += 1
+  return S.region_iter, next
```

Comment:
I'm a bit on the fence about this, but I would lean towards a design where `iterate(::SweepIterator)` actually performs the region iteration as opposed to returning the region iterator, so then `for _ in sweep_iterator end` actually runs the calculation. Then, we could have an iteration adapter such as `region_iterators(::SweepIterator)` returns an iterator that returns the region iterator at each iteration.

Reply:
For reference, in DifferentialEquations.jl, `iterate` is just defined as:
```julia
function Base.iterate(integrator::DEIntegrator, state = 0)
    done(integrator) && return nothing
    state += 1
    step!(integrator)
    return integrator, state
end
```
which seems like a reasonable goal to aim for, and then the complexity of the implementation is in `done(...)` and `step!(...)`.